src/lava/proc/io/source.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+
+import numpy as np
+
+from lava.magma.core.process.process import AbstractProcess
+from lava.magma.core.process.ports.ports import OutPort
+from lava.magma.core.process.variable import Var
+from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
+from lava.magma.core.model.py.ports import PyOutPort
+from lava.magma.core.model.py.type import LavaPyType
+from lava.magma.core.resources import CPU
+from lava.magma.core.decorator import implements, requires, tag
+from lava.magma.core.model.py.model import PyLoihiProcessModel
+
+
+class SendProcess(AbstractProcess):
+    """Spike generator process
+
+    Parameters
+    ----------
+    data: np array
+        data to generate spike from. Last dimension is assumed as time.
+    """
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        data = kwargs.pop('data')
+        self.data = Var(shape=data.shape, init=data)
+        self.s_out = OutPort(shape=data.shape[:-1])  # last dimension is time
+
+
+class AbstractPySendModel(PyLoihiProcessModel):
+    """Template send process model."""
+    s_out = None
+    data = None
+
+    def run_spk(self):
+        buffer = self.data.shape[-1]
+        self.s_out.send(self.data[..., (self.current_ts - 1) % buffer])
+
+
+@implements(proc=SendProcess, protocol=LoihiProtocol)
+@requires(CPU)
+@tag('floating_pt')
+class PySendModelFloat(AbstractPySendModel):
+    """Float send process model."""
+    s_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, float)
+    data: np.ndarray = LavaPyType(np.ndarray, float)
+
+
+@implements(proc=SendProcess, protocol=LoihiProtocol)
+@requires(CPU)
+@tag('fixed_pt')
+class PySendModelFixed(AbstractPySendModel):
+    """Fixed point send process model."""
+    s_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, np.int32, precision=24)
```

Comment:
Somewhere in here, justify and/or document the choice of 24-bit precision for the fixed point model. We are doing it to be compatible with Loihi 2 hardware's graded spikes. But the model is just tagged 'fixed_pt', and has no notion of being bit accurate or anything.

Reply:
Also, as long as you are not using the precision anywhere in your code, there is no real need of specifying it in the LavaPyType.
Otherwise people will ask "what is this for"

Reply:
Actually it would be good to parameterize the precision as an argument to the process definition. But I am not clear how that can be retrieved during Model definition at the moment.

Reply:
Check the builder. I originally put some code in there that would save the precision as an attribute. But currently somebody has commented it out.

As an alternative, you could simply access the LavaPyType:
MyProcClass.some_class_attr.precision

What's actually missing from the LavaPyType is a specification of whether it's signed or unsigned. I could imagine three ways of doing this:
- Introduce another argument for LavaPyType. But adds extra verbosity.
- Introduce negative and positive precision numbers. Negative meaning signed, Positive meaning unsigned. But could be unusual and confusing.
- Using the data type to infer that like np.unit8 or np.int8. Requires that people think consciously about int and uint which is probably ok.